src/graph.c

```diff
@@ -0,0 +1,392 @@
+			todo = explore_add(todo, done, b->sides[i].edge.link);
+		} else {
+			for (int j = 0; j < b->sides[i].max; j++)
+				todo = explore_add(todo, done,
+						   b->sides[i].edges[j].link);
+		}
+	}
+	return todo;
+}
+
+static inline struct pg_brick *get_any_brick(struct pg_graph *graph)
+{
+	struct pg_brick *brick = NULL;
+
+	if (graph->pollable) {
+		brick = graph->pollable->data;
+	} else if (g_hash_table_size(graph->all)) {
+		GHashTableIter iter;
+		gpointer key;
+
+		g_hash_table_iter_init(&iter, graph->all);
+		g_hash_table_iter_next(&iter, &key, (void **)&brick);
+	}
+	return brick;
+}
+
+static inline void empty_graph(struct pg_graph *graph)
+{
+	g_hash_table_steal_all(graph->all);
+	g_slist_free(graph->pollable);
+	graph->pollable = NULL;
+}
+
+int pg_graph_explore_ptr(struct pg_graph *graph,
+			 struct pg_brick *brick,
+			 struct pg_error **error)
+{
+	if (!brick) {
+		brick = get_any_brick(graph);
+		if (!brick) {
+			*error = pg_error_new("Graph %s is empty", graph->name);
+			return -1;
+		}
+	}
+
+	/* first, clear all graph */
+	empty_graph(graph);
+
+	/* explore new graph from brick */
+	GList *todo = NULL;
+	GList *done = NULL;
+
+	todo = g_list_prepend(todo, brick);
+	while (todo != NULL) {
+		struct pg_brick *b = todo->data;
+
+		todo = g_list_remove(todo, b);
+		done = g_list_prepend(done, b);
+		todo = explore_add_neighbours(todo, done, b);
+	}
+
+	/* add all bricks to graph */
+	int ret = 0;
+	GList *h = done;
+
+	while (h) {
+		if (pg_graph_push(graph, h->data, error) < 0) {
+			ret = -1;
+			break;
+		}
+		h = g_list_next(h);
+	}
+	g_list_free(done);
+	return ret;
+}
+
+int pg_graph_poll(struct pg_graph *graph, struct pg_error **error)
+{
+	GSList *n = graph->pollable;
+	uint16_t count;
+
+	while (n) {
+		if (pg_brick_poll(n->data, &count, error) < 0) {
+			if (!*error)
+				*error = pg_error_new("Cannot poll %s",
+					((struct pg_brick *) n->data)->name);
+			return -1;
+		}
+		n = g_slist_next(n);
+	}
+	return 0;
+}
+
+int pg_graph_sanity(struct pg_graph *graph, struct pg_error **error)
+{
+	struct pg_brick *brick = get_any_brick(graph);
+	struct pg_graph *g = pg_graph_new("-", brick);
+	int ret = -1;
+
+	if (!brick)
+		return 0;
+	if (pg_graph_count(g) > pg_graph_count(graph))
+		*error = pg_error_new("Graph not fully explored");
+	else if (pg_graph_count(g) < pg_graph_count(graph))
+		*error = pg_error_new("Graph looks splitted");
+	else
+		ret = 0;
+	empty_graph(g);
+	pg_graph_destroy(g);
+	return ret;
+}
+
+struct pg_brick *pg_graph_get(struct pg_graph *graph, const char *brick_name)
+{
+	struct pg_brick *b;
+
+	if (g_hash_table_lookup_extended(graph->all,
+					 brick_name,
+					 NULL,
+					 (void **) &b))
+		return b;
+	return NULL;
+}
+
+static inline void graph_pop(struct pg_graph *graph, struct pg_brick *b)
+{
+	g_hash_table_remove(graph->all, b->name);
+	if (b->poll)
+		graph->pollable = g_slist_remove(graph->pollable, b);
+}
+
+struct pg_brick *pg_graph_pop(struct pg_graph *graph, const char *brick_name)
```

Comment:
This operation, is unsafe, like really unsafe, because this can easily lead to a broken graph.